### PR TITLE
Refactor the monaco toolbox to be in React, add accessibility support

### PR DESCRIPTION
**Pull request by microsoftsam**
Here's a gif with keystrokes:
monacoaccessibility

Also renaming "Add package" to "Extensions"

### CODE REVIEW

File: webapp/src/blocks.tsx [hunk 1]
```diff
@@ -437,9 +449,8 @@ export class Editor extends srceditor.Editor {
                 if (ev.element == 'category') {
                     let toolboxVisible = !!ev.newValue;
                     this.parent.setState({ hideEditorFloats: toolboxVisible });
-                    if (ev.newValue == lf("{id:category}Add Package")) {
-                        (this.editor as any).toolbox_.clearSelection();
-                        this.parent.addPackage();
+                    if (ev.newValue == lf("{id:category}Extensions")) {
```
Comment (riknoll) [inline webapp/src/blocks.tsx]: Can we move this string to a constant somewhere?

File: webapp/src/monaco.tsx [hunk 2]
```diff
@@ -523,124 +553,157 @@ export class Editor extends srceditor.Editor {
...
+
+            monacoHeadingLabel.appendChild(monacoHeadingIcon);
+            monacoHeadingLabel.appendChild(monacoHeadingText);
+            monacoFlyout.appendChild(monacoHeadingLabel);
         }
 
-        if (this.showAdvanced) {
-            appendCategories(group, namespaces.filter(([, md]) => md.advanced));
+        // Organize and rearrange methods into groups
+        let blockGroups: { [group: string]: MonacoBlockDefinition[] } = {}
```
Comment (riknoll) [inline webapp/src/monaco.tsx]: Use `Map<MonacoBlockDefinition>`

File: webapp/src/monaco.tsx [hunk 3]
```diff
@@ -1271,3 +1202,406 @@ export class Editor extends srceditor.Editor {
...
+
+export class MonacoToolbox extends data.Component<MonacoToolboxProps, MonacoToolboxState> {
+
+    private rootElement: HTMLElement;
+
+    private selectedItem: CategoryItem;
+
+    constructor(props: MonacoToolboxProps) {
+        super(props);
+        this.state = {
```
Comment (riknoll) [inline webapp/src/monaco.tsx]: Use `getInitialState()`

File: webapp/src/monaco.tsx [hunk 4]
```diff
@@ -1271,3 +1202,406 @@ export class Editor extends srceditor.Editor {
...
+
+        let nonAdvancedCategories = createCategories(namespaces.filter(([, md]) => !(md.advanced)));
+        let advancedCategories = showAdvanced && hasAdvanced ? createCategories(namespaces.filter(([, md]) => md.advanced)) : [];
+
+        function stitchCategories(categories: TreeRowProperties[]): TreeRowProperties[] {
+            // Setup previous and next pointers for keyboard navigation between elements
+            for (let i = 0; i < categories.length; i++) {
+                const categoryInfo = categories[i];
+                if (i > 0) {
+                    (categories[i] as any).previousTreeRow = categories[i - 1];;
```
Comment (riknoll) [inline webapp/src/monaco.tsx]: Let's add these two properties to the interface. Just mark it as `/*@internal*/`

File: webapp/src/monaco.tsx [hunk 5]
```diff
@@ -1271,3 +1202,406 @@ export class Editor extends srceditor.Editor {
...
+export interface CategoryItemState {
+    selected?: boolean;
+}
+
+export class CategoryItem extends data.Component<CategoryItemProps, CategoryItemState> {
+    private treeRowElement: TreeRow;
+
+    constructor(props: CategoryItemProps) {
+        super(props);
+        this.state = {
```
Comment (riknoll) [inline webapp/src/monaco.tsx]: Use `getInitialState()`

**Comment by microsoftsam** [comment]
Yes its a big, happy to wait for after the minecraft release but I want it in for adafruit. 
This change also contains making the monaco toolbox blocks have shapes

**Comment by microsoftsam** [comment]
Also this contains the "Extension" rename change. Should I split that out so that it makes the Minecraft release.

**Comment by pelikhan** [comment]
We should get the extension renaming in. Can you factor it out?

**Comment by microsoftsam** [comment]
Done: https://github.com/Microsoft/pxt/pull/3275

**Pull request merged**